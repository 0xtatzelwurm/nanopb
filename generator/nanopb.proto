// Custom options for defining:
// - Maximum size of string/bytes
// - Maximum number of elements in array
//
// These are used by nanopb to generate statically allocable structures
// for memory-limited environments.

import "google/protobuf/descriptor.proto";

enum FieldType {
    FT_DEFAULT = 0; // Automatically decide field type, generate static field if possible.
    FT_CALLBACK = 1; // Always generate a callback field.
    FT_STATIC = 2; // Generate a static field or raise an exception if not possible.
}

message NanoPBOptions {
  optional int32 max_size = 1;
  optional int32 max_count = 2;
  optional FieldType type = 3 [default = FT_DEFAULT];
}

// Protocol Buffers extension number registry
// --------------------------------                                                                                                            
// Project:  Nanopb                                                                                                                            
// Contact:  Petteri Aimonen <jpa@kapsi.fi>                                                                                                    
// Web site: http://kapsi.fi/~jpa/nanopb                                                                                                       
// Extensions: 1010 (all types)                                                                                                                
// --------------------------------   

extend google.protobuf.FileOptions {
    optional NanoPBOptions nanopb_fileopt = 1010;
}

extend google.protobuf.MessageOptions {
    optional NanoPBOptions nanopb_msgopt = 1010;
}

extend google.protobuf.FieldOptions {
    optional NanoPBOptions nanopb = 1010;
}

