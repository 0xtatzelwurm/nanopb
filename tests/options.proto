/* Test nanopb option parsing.
 * options.expected lists the patterns that are searched for in the output.
 */

import "nanopb.proto";

// File level options
option (nanopb_fileopt).max_size = 20;

message Message1
{
    required string filesize = 1;
}

// Message level options
message Message2
{
    option (nanopb_msgopt).max_size = 30;
    required string msgsize = 1;
}

// Field level options
message Message3
{
    required string fieldsize = 1 [(nanopb).max_size = 40];
}


