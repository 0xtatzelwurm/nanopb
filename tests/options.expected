char filesize[20];
char msgsize[30];
char fieldsize[40];
